### ISSUE TITLE
Cover additional cases in implicit object creation analyzer

### ISSUE THREAD

**Issue by mrahhal** [dotnet/roslyn#57789]
There are a few more additional apparent cases that we can cover with this analyzer.

**Languages applicable:**

C# only.

**Code example that the analyzer should report:**

I'm suggesting covering 3 more very common cases:
- Array initializer
```csharp
new C[]
{
    new [|C|](),
    new [|C|](),
};
```
- Simple collection initializer (such as lists, hash sets, etc)
```csharp
new List<C>
{
    new [|C|](),
    new [|C|](),
};
```
- ~Complex collection initializer (such as dictionaries)~
```csharp
new Dictionary<double, C>
{
    { 1, new [|C|]() },
};
```

**Documentation requirements:**

Won't require doc changes (except for maybe adding more example cases).

**Comment by CyrusNajmabadi** [comment]
Pinging @sharwell for next meeting.

**Comment by mavasani** [comment]
These enhancements look reasonable to me.

**Comment by jasonmalinowski** [comment]
In our design meeting, we agreed with the first two examples and support updating the analyzer to support those situations; in both of those cases the "C" is clearly redundant.

We were less convinced by the third example; it's fairly clear in the code snippet that's given there, but consider a case of:

```
new Dictionary<D, C>
{
    { new [|D|](), new [|C|]() },
};
```

In this case you have two different types involved, and even though both of those cases _can_ use implicit new, it's more confusing since there's no longer a single type involved anymore. We felt it's worth leaving this as is _for now_, leaving the proposed behavior change to only apply for the "single argument" initializers and to skip the more complicated case for now.

**Comment by CyrusNajmabadi** [comment]
I'll be helping out with the PR on this.

**Comment by mrahhal** [comment]
> In this case you have two different types involved, and even though both of those cases can use implicit new, it's more confusing since there's no longer a single type involved anymore. We felt it's worth leaving this as is for now, leaving the proposed behavior change to only apply for the "single argument" initializers and to skip the more complicated case for now.

Sounds reasonable. I'll be updating this to remove the 3rd case then.
